Implementing exporting response to file

arnaudelrio opened the pull request:
# Add Export Response Feature and Validation Popups
This PR adds a major usability feature, allowing users to export response data directly from the TUI. The new validation popups provide clear feedback, improving the overall user experience. All changes are backward-compatible and maintain code quality, with no diagnostics issues.

## What's changed
- Implemented **export of HTTP response bodies** to user-specified file paths.
- Introduced **new key binding** (`e`) for exporting responses, configurable in key bindings.
- Added **2 new popups** for export path entry and feedback (success/error). This also gives a way to check that the file was written correctly.

Main logic is implemented in: `src/tui/tui_logic/request/export.rs`. Other files are modified to implement the menu elements, and the keybindings.

---

## Constraints
The filepath must exist and the user must have permissions to write it. It is relative to the path where the TUI was executed. The input must be a valid file path, but doesn't need to exist (it cannot be a folder).

---

## Testing
Works on my ArchLinux system with Bash (Kitty terminal).

I've also tried a few edge cases, like unexisting folders or folders without permissions (gracefully handles the error and displays a popup).

This probably means that ATAC should **not** be used with root permissions in most cases.

---

This would give a "solution" for issues like #186

arnaudelrio committed:
Implementing exporting response to file

arnaudelrio committed:
Fixing code formatting, fixing popups and their color, and implementing exporting for images

arnaudelrio commented:
Hi @Julien-cpsn , can you take a look at this PR when you have time? I am available to discuss any further doubts you may have about this feature, or to correct any issues you find in the code. For instance, I am open to suggestions regarding the keybindings used in this feature.

I still think that this is a very useful feature to have.

I have just updated it to use `edtui` and resolved any merge conflicts with the current upstream.

Julien-cpsn commented:
Hello @arnaudelrio,

Really sorry, I totally forgot about your PR.. let's discuss it!

There are several issues regarding the "philosophy" of this feature, how it's made, and the benefits.

**"Philosophy":**
- ATAC can already export responses in files. In fact, if you set `save_requests_response = true` in your `atac.toml`, the responses will be saved as JSON or YAML format under your collection file. Since every data structure is open-source and public, you can easily interact with it within a rust/python/js script if needed. I get that your proposal is not the exact same thing, but it sounds like a duplicate to me.
- I do not appreciate the usage of generative AI-made code within the TUI runtime. You cannot guarantee working user stories and states when not having the full codebase context, which is the reason why we use Rust. Same for commit reviewing.

**How it's made:**
- The main problem is that your proposal only works with the response's body, thus not including headers, cookies, and js console output.
- The second problem is that it does not provide enough information on what will be done. E.g., will non-existing files and directories be created? Does the file get updated with each response? Does it handle many file formats?
- Yet another popup

**Benefits:**
- My personal opinion is that the benefits are really low regarding how many lines of code are involved. Furthermore, considering my first point, this feature kind of already exists. Especially for something a narrow proportion of users will use.

**My proposal:**
An idea came to my mind when writing these lines, which might seal the deal. I can get that manually parsing the collection files with a third-party script is not that convenient. I propose modifying the `save_requests_response = <bool>` config option by either setting:
- Nothing or "None", wich will not save anything
- "Collection", which will do the same as `true`
- "Separate", which will create a dedicated response file for each request
   - "format", either JSON or YAML (same as `preferred_file_format`)
   - "body", bool
   - "headers", bool
   - "console", bool
   - anything that is pertinent to add

What do you think about it?

Have a great day

arnaudelrio commented:
Hi Julien, thanks for taking a look and for your extended answer!

First off, I would like to state that I also do not like the use of AI for generating code, and this pull request wasn't coded with AI, nor are any of my projects. I'm sorry about the copilot review though, I didn't know it would happen, and I have disabled that settings now.

#### Current implementation and its issues
Back to the feature, I wasn't aware of the existence of this `save_requests_response` flag in the toml file. You are right, it does kinda implement what I wanted. There is no mention of it in the README or in the small documentation. I think it would be a good idea to add an example `atac.toml` file in the repo that contains the default values for every single possible configuration, so that users don't need to search and understand the code in `src/app/files/config.rs`.

The main difference is that I would like to store the response in a fast manner without having to go to the collection file, which is a `json`. The problem is that the response gets escaped, modified, and the images get parsed into numbers. This is not ideal.

I would also like to be able to store the received data in the directory which i executed `atac` from, as it will contain the rest of the data from my back-end (I usually use `atac` to manually edit user's information or similar), and i need the response directly, without any data related to the request (without headers, cookies, etc...). As I would be performing several requests in a row, I thought that being able to input a filename would be helpful.

#### My implementation
Regarding your doubts about edge-case behaviour, I used `std::fs:write()` and the behaviour that goes along with it, so new directories are not created (but `std::fs::created_dir_all();` could be used right before the `std::fs::write()`), if the file doesn't exist it gets created and if it does exist it gets overwritten. **EDIT**: I have just changed it in commit [e31b0cd](https://github.com/Julien-cpsn/ATAC/pull/189/commits/c31b0cd9e585b7b9765ed4ad482b39dc96dc7efc) and it now creates recursively any folders that may not exist using relative paths. Of course, the user must have permission to create the folders.

I am surprised that few people would want this, but I'm sure that being able to copy and paste the response into a file is just as good for most people, only slow for me when I want to execute many requests in a row...

#### Your proposal
Regarding your proposal, I like the idea, and I'm open to changing this. Do you want me to give it a shot? I'm not 100% sure on how to implement it, but I can take a look at it. I also like the idea of being able to export specific parts of the requests, not just the body like I wanted.

Despite this, I am missing a couple of things that I'm not sure your proposal would cover... I would like to:
* Export the response body of _any_ format
* Store the response into a file in the _current working directory_
* Be able to specify a unique _filename_ each time according to the request, so that previous requests are not overwritten

This last issue is my main problem with your proposal, as I'm not sure how it could be done just within the config file. I am trying to come up with a counterproposal, and I think that everything could be simplified from this pull request, except the filename input popup (and consequently the validation one). I can't think of a way to avoid it, besides putting it in the main UI dashboard. What's your take on this?

With all this, I hope you can see why I implemented the pull request this way, although I understand that it adds quite a few lines of code for a feature that is half-implemented. I really do think that we can find a way to make it work for both of us.

Think about it, and let me know what you think!

arnaudelrio committed:
Making sure that the folder exists when exporting the response to a file

arnaudelrio commented:
Hello Julien, I'm guessing you might not have much time lately, but I have been thinking about this, because I've been using my fork, and I think that I have found a way to follow your proposal in a way you'll like.

It might take some time to implement, and I would like your blessing before spending said time. Because of the nature of this new implementation, I would then start fresh in a new PR, would that be ok?

So here's my proposal: similarly to yours, i would not add a popup nor a shortcut but instead I would edit the configuration file. I'd prefer it if this file was editable easily from within the TUI program, but I have to admit that this doesn't sound easy, I'll leave this to someone else.

The difference is that I would like a different kind of flexibility than what you suggested. For example, some users may want to store everything, while users like me would like to be able to use responses directly, or share the collection's configuration without the response being included (in case the response contains confidential information or similar). I don't think there's much use in wanting to store the method, the headers or the cookies only, except for the response body which is usually the target information and this much flexibility would also be quite complicated to implement in my opinion.

I suggest that `save_requests_response` have the following options:
* `none`: doesn't do anything
* `collection` (default): same as current true, saving everything in `yaml` files in the _config folder_
* `request`: saves the collection in the _config folder_, and the request (not collection) without the response in the _current working directory_ with the filename: `<collection>_<requestname>_<method>_request.yaml` (or whatever is fitting)
* `response`: saves the collection in the _config folder_, and the full response in the _current working directory_ as a `yaml` file (overwriting if exists) with the filename: `<collection>_<requestname>_<method>_full.yaml
* `response_body`: saves the collection in the _config folder_, and the response body in the _current working directory_ with the name `<collection>_<requestname>_<method>_body.<extension>`, where the `extension` is obtained according to the MIME type of the response.

I think that this would cover everyone's need, and would make scripting with the responses or sharing request files easier. I'm not sure if I should include combinations of the last three options, for exporting the requests as well as the response, but I don't think it's necessary. As I said before, it would be comfortable to be able to change the configuration from within the TUI. I'm also open to what the filenames should be, because the response body might be a `yaml` too, and it should be distinguishable and avoid overwriting.

Regarding my issues stated in the other comment, it would fully cover the two first points, and I can manage the last issue using a version control system such as `git`, I'll find a way. It even avoids typos and makes it more efficient in a way, if it avoids writing the filename each time, and it makes it more consistent everywhere.

Finally, I think that if this gets implemented, it will important to add it to the documentation properly (which I see that it's totally under construction still).

If you think of something else I'm missing let me know. Thanks for hearing me out!
